modules/common/src/main/scala/com/snowplowanalytics/snowplow/postgres/streaming/Sink.scala

```diff
@@ -12,90 +12,90 @@
 
 import com.snowplowanalytics.iglu.client.Client
 
-import com.snowplowanalytics.snowplow.badrows.{BadRow, Payload, Processor}
+import com.snowplowanalytics.snowplow.badrows.{BadRow, Payload, Processor, Failure}
 import com.snowplowanalytics.snowplow.postgres.api.{DB, State}
 import com.snowplowanalytics.snowplow.postgres.shredding.{Entity, transform}
-import com.snowplowanalytics.snowplow.postgres.streaming.data.{BadData, Data}
+import com.snowplowanalytics.snowplow.postgres.streaming.data.Data
 import com.snowplowanalytics.snowplow.postgres.logging.Slf4jLogHandler
 
-object sink {
+object Sink {
 
   private lazy val logger = getLogger
   private lazy val logHandler = Slf4jLogHandler(logger)
 
   type Insert = ConnectionIO[Unit]
 
+  def sinkResult[F[_]: Sync: ContextShift: Clock: DB](
+    state: State[F],
+    client: Client[F, Json],
+    processor: Processor,
+    badSink: StreamSink[F]
+  )(result: Either[BadRow, Data]): F[Unit] =
+    result.fold(sinkBad[F](_, badSink), sinkGood[F](state, client, processor)(_).flatMap {
+      case Left(badRow) => sinkBad[F](badRow, badSink)
+      case Right(_) => Sync[F].unit
+    })
+
   /**
     * Sink good events into Postgres. During sinking, payloads go through all transformation steps
     * and checking the state of the DB itself.
-    * Events that could not be transformed (due Iglu errors or DB unavailability) are emitted from
-    * the pipe
-    * @param unorderdPipe pipe which might optimise by processing events concurrently
+    * Events that could not be transformed (due Iglu errors or DB unavailability) are emitted as bad data
     * @param state mutable Loader state
     * @param client Iglu Client
     * @param processor The actor processing these events
     */
-  def goodSink[F[_]: Sync: Clock: DB](unorderedPipe: UnorderedPipe[F],
-                                      state: State[F],
-                                      client: Client[F, Json],
-                                      processor: Processor
-  ): Pipe[F, Data, BadData] =
-    unorderedPipe(sinkPayload(state, client, processor)).andThen {
-      _.collect {
-        case Left(badData) => badData
-      }
-    }
-
-  /** Sink bad data coming directly into the `Pipe` */
-  def badSink[F[_]: Sync: ContextShift]: Pipe[F, BadData, Unit] =
-    _.evalMap {
-      case BadData.BadEnriched(row)        => Sync[F].delay(logger.warn(row.compact))
-      case BadData.BadJson(payload, error) => Sync[F].delay(logger.warn(s"Cannot parse $payload. $error"))
-    }
-
-  /** Implementation for [[goodSink]] */
-  def sinkPayload[F[_]: Sync: Clock: DB](state: State[F], client: Client[F, Json], processor: Processor)(
-    payload: Data
-  ): F[Either[BadData, Unit]] = {
+  def sinkGood[F[_]: Sync: Clock: DB](state: State[F], client: Client[F, Json], processor: Processor)(payload: Data): F[Either[BadRow, Unit]] = {
     val result = for {
+      now <- EitherT.liftF(Sync[F].delay(Instant.now))
```

Comment:
I'd still recommend you to use `Clock` for this purpose. If we'll ever want to write a unit test for this function - we can create our own `Clock` instance. Not that I ever did this for functions with `F[_]: Sync: Clock`, but had very good experience with something pure: `F[_]: Monad: Clock`